---
title: "heart_sunitinib_all_concs"
author: "Kejia Li"
date: "2025-03-30"
output:
  html_document:
    keep_md: true
    toc: true
    theme: united
    toc_float: true
    code_folding: hide
  pdf_document:
    toc: true
editor_options:
  
  chunk_output_type: console
---


# 1.General settings 

## 1.1 Chunk settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2, 
  fig.height = 4,
  dpi = 700
)
```


## 1.2 Packages

```{r, message=F, warning=F}

library(progress)
library(tidyverse)
library(ggplot2)
library(reshape2)
library(ggrepel)
library(knitr)
library(ggpubr)
library(ggforce)
library(UniprotR)
library(Peptides)
library(biomaRt)
library(Biostrings)
library(limma)
library(vroom)
library(stringr)
library(corrplot)
library(drc)
library(openxlsx)
library(pheatmap)

mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter
rename <- dplyr::rename

```


## 1.3 Graphics

```{r}

mytheme <- theme_bw()+theme( 
  panel.background = element_rect(color = "black",size=0.5),
  axis.title = element_text(colour="black",size=9),
  axis.text  = element_text(size=8,color="black"),
  legend.position = "right",legend.key.size = unit(0.5,"cm"),
  legend.title = element_blank(),
  legend.text =element_text(colour = "black",size=8)
)

```


# 2. Import data

```{r}

data_dd <- 
  vroom::vroom("sunitinib_dd_heart_report.pr_matrix.tsv") 

colnames(data_dd)

# simplify the colnames
colnames_subset <- colnames(data_dd)[11:42]
common_prefix <- "\\\\msdata1.embl.de\\msdata1\\11_Astral_Stitch\\20250320_Stitch_S4536_LKJ_PELSA_mouse_heart_sunitinib_dd_"
common_prefix <- gsub("\\\\", "\\\\\\\\", common_prefix)
common_suffix <- "_R1.raw"
simplified_colnames <- sub(common_prefix, "", colnames_subset)
simplified_colnames <- sub(paste0(common_suffix, "$"), "", simplified_colnames)  


colnames(data_dd)[11:42] <- simplified_colnames

colnames(data_dd)[c(22:24,38,39)] <- c("5uM_rep1","5uM_rep2","5uM_rep3","control_rep1","control_rep2")

```


# 3 From precursor to peptide

## 3.1 Calculate peptide-level intensity

```{r}

peptide_dd <- data_dd %>%
  group_by(Stripped.Sequence) %>%
  mutate(across(c("1uM_rep1":"02nM_rep4"), ~sum(.x, na.rm = TRUE), .names = "sum_{col}")) %>% 
  distinct(Stripped.Sequence, .keep_all = TRUE) %>%
  ungroup() 

```


## 3.2 Annotate peptide with location

```{r}

diann_prec_add_peptide_pos2 <- function(data_precursor, fasta, silent = T) {
  
  if (!hasArg(data_precursor)) stop("Please specify a data frame or a tibble of the DIA-NN precursors.")
  
  if (!hasArg(fasta)) fasta <- choose.files(multi = F, )
  fasta <- Biostrings::readAAStringSet(fasta)
  fasta <- as.character(fasta)
  names(fasta) <- unlist(lapply(strsplit(x = names(fasta), "\\|"), \(x) x[2]))
  
  if (!silent) {
    message(paste0(nrow(data_precursor), " precursors identified. This should take around ", 
                   round(nrow(data_precursor) * 0.0000884 + 0.1463016, 2), " s."))
    t0 <- Sys.time()
  }
  
  # Add peptide positions
  data_pos <- data_precursor %>% 
    dplyr::rowwise() %>% 
    dplyr::mutate(peptide.length = nchar(Stripped.Sequence), 
                  from = regexpr(Stripped.Sequence, as.character(fasta[unlist(strsplit(Protein.Group, ";"))[1]]))[1], 
                  to = from + peptide.length - 1, 
                  .after = Modified.Sequence) %>% 
    dplyr::ungroup()
  
  if (!silent) {
    t1 <- Sys.time()
    
    duration <- t1 - t0
    
    if (duration > nrow(data_precursor) * 0.0000884 + 0.1463016) {
      message(paste0("Sorry for the delay! It took ", round(duration, 2), " s."))
    } else {
      message(paste0("Wow, even faster! It only took ", round(duration, 2), " s."))
    }
  }
  
  return(data_pos)
  
}

# Input the imported precursor data frame and specify the fasta file
peptide_dd_with_position <- diann_prec_add_peptide_pos2(data_precursor = peptide_dd, fasta = "Mus_musculus_uniprotkb_proteome_UP000000589_AND_revi_2023_09_25.fasta", silent = F)

save(peptide_dd_with_position,file = "peptide_dd_with_position.Rdata")

```


## 3.2 Extract different replicates

```{r}

colnames(peptide_dd_with_position)

peptide_dd_rep1 <- peptide_dd_with_position %>% 
  select(1:13, contains("sum") & contains("rep1"))%>%
  filter(rowSums(.[14:21]==0) == 0) 


peptide_dd_rep2 <- peptide_dd_with_position %>% 
  select(1:13, contains("sum") & contains("rep2")) %>%
  filter(rowSums(.[14:21]==0) == 0) 

peptide_dd_rep3 <- peptide_dd_with_position %>% 
  select(1:13, contains("sum") & contains("rep3")) %>%
  filter(rowSums(.[14:21]==0) == 0) 

peptide_dd_rep4 <- peptide_dd_with_position %>% 
  select(1:13, contains("sum") & contains("rep4")) %>%
  filter(rowSums(.[14:21]==0) == 0) 

```


# 3.3 visualize data completeness

```{r}

colnames(peptide_dd_with_position)

peptide_dd_final <- peptide_dd_with_position %>%
  mutate(
    rep1 = ifelse(Stripped.Sequence %in% peptide_dd_rep1$Stripped.Sequence, 1, 0),  
    rep2 = ifelse(Stripped.Sequence %in% peptide_dd_rep2$Stripped.Sequence, 1, 0),  
    rep3 = ifelse(Stripped.Sequence %in% peptide_dd_rep3$Stripped.Sequence, 1, 0),  
    rep4 = ifelse(Stripped.Sequence %in% peptide_dd_rep4$Stripped.Sequence, 1, 0), 
    total_count = rep1 + rep2 + rep3 + rep4
  ) %>% select(1:11,46:82)

colnames(peptide_dd_final)


peptide_dd_final$total_count <- as.factor(peptide_dd_final$total_count)
reps_count <- as.data.frame(table(peptide_dd_final$total_count))
colnames(reps_count) <- c("number_of_reps", "count")
colnames(peptide_dd_final)


ggplot(data=reps_count ,mapping=aes(x= number_of_reps, y=count)) + 
  geom_bar(stat = "identity",position = "dodge",color="black")+
  theme_classic()+
  geom_text(aes(label = count), vjust = -0.5, color = "black", size = 4) 

```


# 3.4 calculate peptide affinity 

```{r}

# Only run this code if fitting data is not present (takes ~2 h per replicate)

if (!file.exists("export_dd_suniti_peptide_stabilized_R1.csv")) {
  
  colnames(peptide_dd_rep1)
  
  # Calculate log2FCå’ŒRatio
  calculate_ratios <- function(data, rep) {
    for (i in c("control","1uM","02nM","2nM","5uM", "20nM","20uM","200nM")) {
      data[, paste0("log2FC_", i, "_0uM_", rep)] <- log2(data[, paste0("sum_", i, "_", rep)] / data[, paste0("sum_control_", rep)])
      data[, paste0("Ratio_", i, "_control_", rep)] <- data[, paste0("sum_", i, "_", rep)] / data[, paste0("sum_control_", rep)]
    }
    return(data)
  }
  
  
  colnames(peptide_dd_rep1)
  peptide_dd_rep1 <- calculate_ratios(peptide_dd_rep1, "rep1")
  peptide_dd_rep2 <- calculate_ratios(peptide_dd_rep2, "rep2")
  peptide_dd_rep3 <- calculate_ratios(peptide_dd_rep3, "rep3")
  peptide_dd_rep4 <- calculate_ratios(peptide_dd_rep4, "rep4")
  
  
  save(peptide_dd_rep1,file = "peptide_dd_rep1.Rdata")
  save(peptide_dd_rep2,file = "peptide_dd_rep2.Rdata")
  save(peptide_dd_rep3,file = "peptide_dd_rep3.Rdata")
  save(peptide_dd_rep4,file = "peptide_dd_rep4.Rdata")
  
  
  # pre-filter peptides
  subset_target <- peptide_dd_rep1 %>%
    filter(Ratio_20uM_control_rep1< 0.7 & Ratio_5uM_control_rep1 < 1)%>%
    select(Stripped.Sequence,Genes,First.Protein.Description,Protein.Ids,contains("Ratio"))%>% gather(concentration,Ratio,paste0("Ratio_","control","_control_rep3"):paste0("Ratio_","20uM","_control_rep1")) 
  
  
  ## pre-filter peptides (destabilized)
  # subset_target <- peptide_dd_rep %>%
  #   filter(Ratio_20uM_control_rep1 > 1.3 & Ratio_5uM_control_rep1 > 1)%>%
  #   select(Stripped.Sequence,Genes,First.Protein.Description,Protein.Ids,contains("Ratio"))%>% gather(concentration,Ratio,paste0("Ratio_","control","_control_rep3"):paste0("Ratio_","20uM","_control_rep1")) 
  
  
  # fit the curve
  subset_target$EC50<-0
  subset_target$R2<-0
  subset_target$lower <- 0
  subset_target$upper <- 0
  EC50.col <- grep("EC50",colnames(subset_target))
  R2.col <- grep("R2",colnames(subset_target))
  lower.col <- grep("lower",colnames(subset_target))
  upper.col <- grep("upper",colnames(subset_target))
  
  data_peptide <- data.frame(conc=c(log10(1e-12),log10(2e-10),log10(2e-9),log10(2e-8),log10(2e-7),log10(2e-6),log10(2e-5),log10(1e-4)), ratio=subset_target$Ratio)
  
  
  pb <- progress_bar$new(
    format = "[:bar] :percent :elapsed",
    total = length(subset_target$Stripped.Sequence), 
    clear = FALSE, 
    width = 60 
  )
  
  state <- c(0)
  for (peptide in subset_target$Stripped.Sequence) {
    pb$tick()
    state <- state + 1
    #*****
    data_peptide <- data.frame(conc=c(log10(0),log10(2e-10),log10(2e-9),log10(2e-8),log10(2e-7),log10(2e-6),log10(2e-5),log10(1e-4)),ratio=subset_target[subset_target$Stripped.Sequence==peptide,]$Ratio)
    
    peptide_drm <- tryCatch(
      drm(ratio~conc,data=data_peptide,fct = LL.4(),logDose = 10),
      error=function(e){ print("Correlation LL.4 failed");return(NA)}
    )
    
    
    # export PEC50
    subset_target[state,EC50.col] <- tryCatch(
      round(-log(coef(peptide_drm)[[4]],base =10),digits = 4),
      error=function(e){return(NA)}
    )
    
    # export lower
    subset_target[state,lower.col] <- tryCatch(
      as.numeric(coef(peptide_drm)[2]),
      error=function(e){return(NA)}
    )
    
    # export upper
    subset_target[state,upper.col] <- tryCatch(
      as.numeric(coef(peptide_drm)[3]),
      error=function(e){return(NA)}
    )
    
    
    # export R
    subset_target[state,R2.col] <- tryCatch(
      round((cor(data_peptide$ratio, fitted(peptide_drm), method = "pearson"))^2,digits = 4),
      error=function(e){return(NA)}
    )
    
  }
  
  stabilized_peptides <- subset_target %>%
    pivot_wider(names_from = concentration, values_from = Ratio)
  class(stabilized_peptides)
  
  # output
  write.csv (file = paste0("export_dd_suniti_peptide_stabilized_R1",".csv"), 
             stabilized_peptides, row.names = F)
  
}

```


# 4 Combine affinity data

## 4.1 affinity data import

```{r}

affinity_rep1 <- 
  bind_rows(
    "stabilized" =  read.csv("export_dd_suniti_peptide_stabilized_R1.csv"),
    "destabilized"= 
      read.csv("export_dd_suniti_peptide_destabilized_R1.csv"),
    .id = "type") %>%
  filter(R2 >0.9) %>%
  rename(EC50_rep1= EC50, R2_rep1=R2,lower_rep1 = lower, upper_rep1=upper)



affinity_rep2 <- 
  bind_rows(
    "stabilized" =  read.csv("export_dd_suniti_peptide_stabilized_R2.csv"),
    "destabilized"= 
      read.csv("export_dd_suniti_peptide_destabilized_R2.csv"),
    .id="type") %>%
  filter(R2 >0.9) %>%
  rename(EC50_rep2= EC50, R2_rep2=R2,lower_rep2 = lower, upper_rep2=upper)


affinity_rep3 <- 
  bind_rows(
    "stabilized" =  read.csv("export_dd_suniti_peptide_stabilized_R3.csv"),
    "destabilized"= 
      read.csv("export_dd_suniti_peptide_destabilized_R3.csv"),
    .id="type") %>%
  filter(R2 >0.9)%>%
  rename(EC50_rep3= EC50, R2_rep3=R2,lower_rep3 = lower, upper_rep3=upper)

affinity_rep4 <- 
  bind_rows(
    "stabilized" =  read.csv("export_dd_suniti_peptide_stabilized_R4.csv"),
    "destabilized"= 
      read.csv("export_dd_suniti_peptide_destabilized_R4.csv"),
    .id="type") %>%
  filter(R2 >0.9)%>%
  rename(EC50_rep4= EC50, R2_rep4=R2,lower_rep4 = lower, upper_rep4=upper)

```


## 4.2 add affinities to each peptide

```{r}

affinity_reps <- list(affinity_rep1, affinity_rep2, affinity_rep3, affinity_rep4)


sta_list <- lapply(affinity_reps, function(x) filter(x, type == "stabilized"))
desta_list <- lapply(affinity_reps, function(x) filter(x, type == "destabilized"))



peptide_dd_affinity_count <- peptide_dd_final %>%
  mutate(
    # stabilized counts
    rep1_sta_dd = ifelse(Stripped.Sequence %in% sta_list[[1]]$Stripped.Sequence, 1, 0),
    rep2_sta_dd = ifelse(Stripped.Sequence %in% sta_list[[2]]$Stripped.Sequence, 1, 0),
    rep3_sta_dd = ifelse(Stripped.Sequence %in% sta_list[[3]]$Stripped.Sequence, 1, 0),
    rep4_sta_dd = ifelse(Stripped.Sequence %in% sta_list[[4]]$Stripped.Sequence, 1, 0),
    total_dd_count_sta = rep1_sta_dd +rep2_sta_dd+ rep3_sta_dd +rep4_sta_dd,
    
    # destabilized counts
    rep1_des_dd = ifelse(Stripped.Sequence %in% desta_list[[1]]$Stripped.Sequence, 1, 0),
    rep2_des_dd = ifelse(Stripped.Sequence %in% desta_list[[2]]$Stripped.Sequence, 1, 0),
    rep3_des_dd = ifelse(Stripped.Sequence %in% desta_list[[3]]$Stripped.Sequence, 1, 0),
    rep4_des_dd = ifelse(Stripped.Sequence %in% desta_list[[4]]$Stripped.Sequence, 1, 0),
    total_dd_count_des = rep1_des_dd +rep2_des_dd+ rep3_des_dd + rep4_des_dd)  %>%
  mutate(across(c(total_dd_count_sta, total_dd_count_des),as.factor))  %>%
  mutate(regulation = 
           case_when(
             as.numeric(as.character(total_dd_count_des)) >2 ~"destabilized",
             as.numeric(as.character(
               total_dd_count_sta)) > 2 ~"stabilized",
             as.numeric(as.character(
               total_count)) <3 ~"underquantified",
             as.numeric(as.character(
               total_count)) >2 ~"unresponsive",
             TRUE~"unknown"))


reps_count <- as.data.frame(table(peptide_dd_affinity_count$total_dd_count_sta))

colnames(reps_count) <- c("number_of_reps", "count")


ggplot(data=reps_count ,mapping=aes(x= number_of_reps, y=count))+geom_bar(stat = "identity",position = "dodge",color="black")+
  theme_classic()+
  geom_text(aes(label = count), vjust = -0.5, color = "black", size = 4) 


combine <- peptide_dd_affinity_count %>%
  left_join(sta_list[[1]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids"))%>%
  left_join(sta_list[[2]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids"))%>%
  left_join(sta_list[[3]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids"))%>%
  left_join(sta_list[[4]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids")) %>%   left_join(desta_list[[1]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids"))%>%
  left_join(desta_list[[2]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids"))%>%
  left_join(desta_list[[3]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids"))%>%
  left_join(desta_list[[4]],by=c("Stripped.Sequence","Genes","First.Protein.Description","Protein.Ids")) %>%
  select(-contains("type"))

colnames(combine)

colnames_all <- names(combine[c(61:155)])
dupe_prefixes <- unique(gsub("\\.(x|y)$", "", colnames_all))

for (col in dupe_prefixes) {
  col_x <- paste0(col, ".x")
  col_y <- paste0(col, ".y")
  
  combine[[col]] <- coalesce(combine[[col_x]], combine[[col_y]])
  
  combine[[col_x]] <- NULL
  combine[[col_y]] <- NULL
}


calculate_cv_all <- function(x) {
  valid_values <- na.omit(x)
  (sd(valid_values) / mean(valid_values))}


final_list <- combine %>%
  mutate(
    CV_EC50 = NA_real_,
    mean_EC50 = NA_real_,
    
    CV_EC50 = ifelse(regulation %in% c("stabilized", "destabilized"),
                     apply(select(., starts_with("EC50_")), 1, calculate_cv_all),
                     CV_EC50),
    
    mean_EC50 = ifelse(regulation %in% c("stabilized", "destabilized"),
                       rowMeans(across(starts_with("EC50_rep")), na.rm = TRUE),
                       mean_EC50),
    is.EC50.reproducible = 
      case_when(
        CV_EC50 < 0.25 ~"yes",
        CV_EC50 > 0.25 ~"no",
        TRUE ~"unknown"))


write.xlsx(final_list, file = "Supplementary.Dataset8.csv")

```


# 5.Fig.3H: Heatmap showing the targets of sunitinib in mouse heart

## 5.1 Data preparation

```{r}

colnames(final_list)
table(final_list$regulation)

dose_response_peptides <- final_list %>% 
  filter(regulation=="stabilized")     # no destabilized peptides are observed


## this is for annotation
other_columns <- dose_response_peptides %>%
  filter(CV_EC50 < 0.25) %>%
  select(-contains("_rep"))


data_plot <- dose_response_peptides %>%
  filter(CV_EC50 < 0.25) %>%
  select(1:11, contains("Ratio")) %>% 
  pivot_longer(
    cols = contains("_rep"),
    names_to = c(".value", "reps"),
    names_sep = "_rep") %>%
  group_by(Stripped.Sequence) %>%
  summarise(across(contains("Ratio"), mean, na.rm = TRUE)) %>%
  left_join(other_columns,by="Stripped.Sequence")



kinase <- read.csv("mouse_kinase_unprot.csv", 
                   header = TRUE)


protein_level <- data_plot %>%
  arrange(desc(mean_EC50)) %>%
  distinct(Protein.Group,.keep_all = TRUE) %>%
  mutate(across(contains("Ratio"), 
                ~ log2(.), 
                .names = "{.col}_log2")) %>%
  mutate(is.kinase= ifelse(Protein.Group %in% kinase$Entry,"yes","no"))


write.csv(file = "SourceData_Fig3G.csv", protein_level, row.names = F)

```


## 5.2 Plotting

```{r}

ann_colors <- list(mean_EC50 = colorRampPalette(c("#f0b38c", "#6c1f58"))(10),
                   is.kinase = c("no" = "#cfcfcf","yes" = "#1d6375"))

annotation_is.kinase  <- protein_level %>%
  select(13,47) %>%
  tibble::column_to_rownames(var = names(.)[1]) 

annotation_EC50 <- protein_level %>%
  select(13,37)%>%
  tibble::column_to_rownames(var = names(.)[1]) 



plot <- protein_level %>%
  select(contains("_log2"))

colnames(plot) <- c("control","0.2 nM","2 nM","20 nM",
                    "200 nM","1 uM","5 uM","20 uM")

plot <- as.matrix(plot)

row.names(plot) <- protein_level$Genes



my_colors <- colorRampPalette(c("#176226","#f7fbff"))(40)

pdf(file = "Fig.3H.pdf",width = 8,height = 8,onefile = FALSE)

pheatmap(plot,cluster_rows = F, cluster_cols = F,
         annotation_names_col = F,
         annotation_names_row = F,
         annotation_row = cbind(annotation_EC50,annotation_is.kinase),
         show_rownames = TRUE, 
         color = my_colors,
         annotation_colors = ann_colors) 
dev.off()

dev.list()

while (dev.cur() > 1) dev.off() 

```
