---
title: "crude_liver"
author: "kejia"
date: "2025-03-03"
output:
  html_document:
    keep_md: true
    toc: true
    theme: united
    toc_float: true
    code_folding: hide
  pdf_document:
    toc: true
editor_options:
  
  chunk_output_type: console
---

# 1.General settings 
## 1.1 Chunk settings
```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2, 
  fig.height = 4,
  dpi = 700
)
```

## 1.2 Packages

```{r, message=F, warning =F}
library(tidyverse)
library(ggplot2)
library(reshape2)
library(ggrepel)
library(knitr)
library(ggpubr)
library(ggforce)
library(UniprotR)
library(Peptides)
library(biomaRt)
library(Biostrings)
library(limma)
library(vroom)
library(stringr)
library(corrplot)
library(ggh4x)
library(openxlsx)
library(htmltools)


mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter
rename <- dplyr::rename

```


# 2. Import data
```{r}

liver <- vroom::vroom("staurosporine_liver_report.pr_matrix.tsv")


diann_prec_add_peptide_pos2 <- function(data_precursor, fasta, silent = T) {
  
  if (!hasArg(data_precursor)) stop("Please specify a data frame or a tibble of the DIA-NN precursors.")
  
  if (!hasArg(fasta)) fasta <- choose.files(multi = F, )
  fasta <- Biostrings::readAAStringSet(fasta)
  fasta <- as.character(fasta)
  names(fasta) <- unlist(lapply(strsplit(x = names(fasta), "\\|"), \(x) x[2]))
  
  if (!silent) {
    message(paste0(nrow(data_precursor), " precursors identified. This should take around ", 
                   round(nrow(data_precursor) * 0.0000884 + 0.1463016, 2), " s."))
    t0 <- Sys.time()
  }
  
  # Add peptide positions
  data_pos <- data_precursor %>% 
    dplyr::rowwise() %>% 
    dplyr::mutate(peptide.length = nchar(Stripped.Sequence), 
                  from = regexpr(Stripped.Sequence, as.character(fasta[unlist(strsplit(Protein.Group, ";"))[1]]))[1], 
                  to = from + peptide.length - 1, 
                  .after = Modified.Sequence) %>% 
    dplyr::ungroup()
  
  if (!silent) {
    t1 <- Sys.time()
    
    duration <- t1 - t0
    
    if (duration > nrow(data_precursor) * 0.0000884 + 0.1463016) {
      message(paste0("Sorry for the delay! It took ", round(duration, 2), " s."))
    } else {
      message(paste0("Wow, even faster! It only took ", round(duration, 2), " s."))
    }
  }
  
  return(data_pos)
  
}


# Input the imported precursor data frame and specify the fasta file
fasta_path <- "Mus_musculus_uniprotkb_proteome_UP000000589_AND_revi_2023_09_25.fasta"

Annotated_liver <- diann_prec_add_peptide_pos2(liver, fasta_path, F)

```

# 3. Divide into two groups

```{r}

colnames(Annotated_liver)

renamed_liver<-  Annotated_liver %>%
  select(1:21) %>%
  rename(
    control_1 = 18, control_2 = 19, control_3 = 20,
    control_4 = 21, drug_1 = 14, drug_2 = 15,
    drug_3 = 16, drug_4 = 17
  )



```


# 4. Statistic test analysis

```{r, include=FALSE,eval=FALSE}

analyze_dataset <- function(data, dataset_name) {
  
  
  # step-1：data-processing
  peptide_intensity <- data %>%
    group_by(Stripped.Sequence) %>%
    mutate(across(c(drug_1, drug_2, drug_3, drug_4, control_1, control_2, control_3, control_4), ~sum(.x, na.rm = TRUE), .names = "sum_{col}")) %>%
    select(Protein.Ids, Genes, First.Protein.Description, Stripped.Sequence, peptide.length, from, to, 
           starts_with("sum_")) %>%
    distinct(Stripped.Sequence, .keep_all = TRUE) %>%
    ungroup() %>%
    filter(rowSums(.[8:15] == 0) == 0)
  
  
  # step-2：Bayes t-test
  data <- peptide_intensity %>% select(c(1:7, 12:15, 8:11))
  columnnames <- c("Protein.Ids", "Genes", "name", "sequence", "length", "start", "end", 
                   "control_1", "control_2", "control_3", "control_4", "D1_1", "D1_2", "D1_3", "D1_4", "UID")
  design <- model.matrix(~ 0 + factor(c(0, 0, 0, 0, 1, 1, 1, 1))) # 0=Vehicle; 1=Drug
  colnames(design) <- c("V", "D") # 0=Vehicle; 1=Drug
  
  data.DIA <- grep("_", colnames(data))
  data[, data.DIA] <- log((data[, data.DIA]), base = 2)
  data.metanms <- grep("Stripped.Sequence", colnames(data))
  data$last <- (regexpr(';', as.character(data$Protein.Ids)) - 1) # if no ";" then = -1
  data$UID <- ifelse(data$last > 0, (substr(as.character(data$Protein.Ids), 1, data$last)),
                     (as.character(data$Protein.Ids)))
  data$last <- NULL
  intensity <- grep("_", colnames(data))
  data <- setNames(data, columnnames)
  data$ID <- 1:nrow(data)
  
  fit <- lmFit(data[, intensity], design)
  fit <- eBayes(fit) ## Apply empirical Bayes smoothing to the standard errors.
  contrast <- makeContrasts("D-V", levels = design)
  fit2 <- contrasts.fit(fit, contrast)
  fit2 <- eBayes(fit2, trend = TRUE)
  
  ptopf <- topTableF(fit2, adjust = "BH", genelist = data[, c("ID")], number = Inf)
  ptopf$rank <- 1:length(ptopf$F)
  ptopf$numfp <- ptopf$rank * ptopf$adj.P.Val
  
  t.stat <- topTable(fit2, adjust = "BH", genelist = data[, "ID"], coef = 1, p.value = 1, number = Inf)
  nms <- grep("logFC", colnames(t.stat)) 
  colnames(t.stat)[nms] <- "Log2FC"
  t.stat$Log10P.Value <- -log(t.stat$P.Value, base = 10)
  t.stat$Log10adj.P.Val <- -log(t.stat$adj.P.Val, base = 10)
  t.statNMS <- t.stat[c("ID", "Log2FC", "Log10P.Value", "Log10adj.P.Val", "P.Value", "adj.P.Val", "AveExpr", "t", "B")]
  t.stat <- t.statNMS
  
  data.analysis <- merge(t.stat, data, by = "ID", all = TRUE, sort = FALSE)
  
  
  
  # write csv or xlsx
  write.csv(data.analysis, file = paste0("bayes_test_",dataset_name,".csv"))
  
  
}



datasets <- list(
  liver_staurosporine = renamed_liver)

library(purrr)
map2(datasets, names(datasets), ~ analyze_dataset(.x, .y))


```


# 4.Data analysis

## 4.1 Fig.3E: Volcano plot showing staurosporine target identification in the mouse liver lysates
```{r}

kinase <- read.csv("mouse_kinase_unprot.csv")

data_for_volcano <- read.csv("bayes_test_liver_staurosporine.csv") %>% 
  mutate(is.kinase = ifelse(UID %in% kinase$Entry,"yes","no")) %>% 
  distinct(UID,.keep_all = TRUE)

p <- ggplot(data = data_for_volcano , aes(x = Log2FC, y = Log10P.Value, fill = is.kinase)) +
  geom_point(shape = 21,color = "black", size = 2, alpha = 0.7,stroke= 0.2) +  
  scale_fill_manual(name= "is.kinase",breaks = c("yes","no"), values = c("#E69F00","#b3b3b3")) + scale_x_continuous(
    name = expression(Log["2"] * FC),
    limits = c(-5, 5),
    breaks = -4:4
  ) +
  scale_y_continuous(
    name = expression(-Log["10"] * Pvalue),
    limits = c(0, 14),
    breaks = seq(0, 14, by = 2)
  ) +
  labs(
    title = "liver_staurosporine",
    color = "is.kinase"  
  ) +
  annotate(
    "text", 
    x = 0, y = 13, 
    label = paste0("Protein number = ", nrow(data_for_volcano)), 
    size = 5, 
    hjust = 0.5, 
    vjust = 1.5, 
    color = "black"
  ) +
  theme_minimal()+ 
  theme(
    plot.title = element_text(size = 14, face = "bold", hjust = 0.5), 
    axis.title = element_text(size = 12), 
    axis.text = element_text(size = 12,color="black"), 
    legend.title = element_text(size = 12), 
    legend.text = element_text(size = 10),  
    panel.grid.major = element_line(color = "grey80", linewidth = 0.2),  
    panel.grid.minor = element_blank(),  
    panel.background = element_rect(fill = "white", color = "black", linewidth = 1),  
    plot.background = element_rect(fill = "white",colour = NA),axis.ticks = element_line(color = "black", linewidth = 0.5),  
    axis.ticks.length = unit(0.2, "cm")  
  )

p


ggsave(paste0("Fig.3E",".pdf"),width = 6,height = 5.5,useDingbats=FALSE)

```


## 4.2 Fig.3F: Barplot to compare the performance of HT-PELSA and PISA in identifying staurosporine targets in the liver lysates
```{r}

htPELSA_target <- read.csv("bayes_test_liver_staurosporine.csv") %>% 
  mutate(is.kinase = ifelse(UID %in% kinase$Entry,"yes","no")) %>% 
  distinct(UID,.keep_all = TRUE) %>% 
  filter(Log2FC <0) %>% 
  head(122) %>% 
  group_by(is.kinase) %>% 
  count()


## this list is downloaded from literature "Nature Communications volume 15, Article number: 8923 (2024)"

PISA <- read.xlsx("PISA_tissue.xlsx", sheet = "Staurosporin - Rat Liver")

PISA_target <- PISA %>% 
  mutate(pass = if_any(156:161, ~ .x == TRUE)) %>% 
  filter(pass== TRUE) %>% 
  group_by(isKinase) %>% 
  count() %>% 
  rename(is.kinase=isKinase) %>% 
  mutate(is.kinase = case_when(
    is.kinase == TRUE ~ "yes",
    is.kinase == FALSE ~ "no",
    TRUE ~ as.character(is.kinase)  
  ))



data_plot <- rbind(
  "htPELSA" = htPELSA_target,
  "PISA" = PISA_target,
  .id="type")%>% 
  mutate(is.kinase = case_when(
    is.kinase == "yes" ~ "Kinase",
    is.kinase == "no" ~ "non-Kinase",
    TRUE ~ as.character(is.kinase)  
  )) 




p <- ggplot(data = data_plot ) +
  geom_bar(
    aes(x = type, y = n, fill = factor(is.kinase, levels = c("non-Kinase", "Kinase"))),
    stat = "identity", width = 0.4, position = "stack"
  ) +
  scale_y_continuous(
    "Identifications",
    breaks = c(0, 50,100,150,200,250,300),
    labels = c(0, 50,100,150,200,250,300), 
    limits = c(0, 320))+ 
  scale_fill_manual(
    name = "",
    breaks = c("non-Kinase", "Kinase"),
    labels = c("non-Kinase", "Kinase"),
    values = c("gray75", "#a995c4")
  ) + theme_classic() + xlab(NULL) + ylab(NULL)+
  annotate(x=rep(c("PISA", "htPELSA"),times=2),y=c(6,40,74,110),label=c("10","90","302","32"),geom = "text",size=2)+
  theme(legend.position = "top")
p


ggsave("Fig3F.pdf",width =8,height =10 ,units = "cm")

```

